## /users/abrick/pr/231/2647/vj.py

In [3]:
# 05 Time
#
#
# 05 Time
#
# Demonstrate the use of a generator that indicates
# for each event in access_log the number of seconds
# elapsed since the most recent midnight

# A generator the provides the number of seconds since midnight
# for each event in the log
def get_seconds():
  log = 'access_log'
  for line in open( log ):
    # the timestamp is the string between the square brackets [...]
    timestamp = line.split('[',1)[1].split(']',1)[0]

    # remove the date and timezone from the timestamp
    timestamp = timestamp.split()[0]
    timestamp = timestamp.split(':')

    # convert the hours, minutes, and seconds, in the the seconds since midnight
    seconds = int(timestamp[1])*3600 + int(timestamp[2])*60 + int(timestamp[3])
    yield seconds


# create the generator
elapsed_seconds = get_seconds()

# get the next value from the generator and print it
for seconds in elapsed_seconds:
  print( seconds )

34560
34560
34561
34696
34697
34698
34739
34947
34948
34948
34948
34948
34948
34948
34949
34949
34949
34949
35052
35054
35055
35192
35192
35192
35192
35192
35192
35192
35193
35193
35193
35193
35196
35416
35416
35417
35532
35533
35533
35533
35533
35533
35533
35533
35533
35534
35534
35553
35556
35557
35634
35635
35638
35915
35984
35985
35985
35985
35985
35985
35985
35985
35985
35985
35985
36093
36093
36093
36093
36093
36093
36093
36093
36093
36093
36093
36093
36093
36093
36093
36189
36189
36196
36205
36206
36206
36206
36206
36206
36206
36206
36206
36206
36206
36209
36209
36219
36229
36258
36259
36260
36284
36291
36291
36437
36604
36625
36763
36829
36835
36836
36837
36848
36848
36848
36848
36848
36848
36848
36848
36848
36848
36848
36848
36848
36849
36859
36860
36861
36993
36995
36996
36998
37021
37022
37136
37181
37239
37241
37571
37571
37571
37571
37572
37572
37572
37572
37572
37572
37572
37572
37572
37572
37819
37996
38037
38037
38037
38037
38037
38038
38038
38038
38038
38038
38038
3815

20804
20810
20831
21060
21061
21085
21152
21157
21158
21263
21262
21608
21611
21938
21940
21963
22074
22370
22398
22478
22689
22727
22735
23034
23038
23065
23080
23082
23088
23089
23090
23092
23242
23245
23477
23907
23942
24216
24435
24460
24466
24481
24490
24493
24495
24496
24496
24500
24580
24644
24723
24748
24815
25187
25200
25205
25207
25344
25435
25507
25582
25643
25786
25787
26025
26119
26166
26180
26180
26249
26480
26544
26652
26716
26769
26858
26870
27111
27588
27779
27786
27786
27812
27860
27861
27923
27924
27924
27924
27924
27924
27924
27924
27925
27925
27925
27925
28055
28123
28129
28130
28160
28179
28857
28898
28912
28912
28964
29118
29120
29122
29148
29196
29318
29319
29320
29471
29663
29666
29966
29968
29969
30018
30025
30030
30249
30251
30292
30298
30483
30574
30795
30795
30796
30796
31097
31262
31281
31284
31295
31475
31502
31664
31806
31808
31815
31843
31845
31903
31944
31957
31962
31968
31973
31973
31978
31978
31981
31983
31984
31985
31989
32037
32170
32228
32229
3223

KeyboardInterrupt: 

## /users/abrick/pr/231/2647/33.py

In [35]:
# #Homework  6
# Demonstrate the use of a generator that indicates for each event in access_log
# the number of seconds elapsed since the most recent midnight
from datetime import datetime
from datetime import date
import re

today = date.today()

midnight = datetime(today.year,today.month, today.day)


time_stamp =(re.findall('\d{2}/\w{3}/\d{4}\:\d\d\:\d\d\:\d\d', line)[0]
             for line in open('access_log')) # creating a list of all dates that found in the text
after_midnight = (((datetime.strptime(x, '%d/%b/%Y:%H:%M:%S') - midnight).days*24*3600+(datetime.strptime(x, '%d/%b/%Y:%H:%M:%S')- midnight).seconds) #converts string to date
                  #check that dates bigger than midnight
                  #put it in the list "after_midnight"
                  for x in time_stamp if datetime.strptime(x, '%d/%b/%Y:%H:%M:%S') >= midnight)
#created a list of 10 elements from "after_midnight" list and print it
for i in [next(after_midnight) for _ in range(10)]:
    print (i)

370
370
370
566
566
566
669
765
765
765


Peer review: 

At first I thought I wasn't getting any output, because it has to parse through the entire list to meet this condition: "if datetime.strptime(x, '%d/%b/%Y:%H:%M:%S') >= midnight". This was one interpretation of the prompt, so obviously won't dock you for that interpration; but the discussion thread eventually said it was seconds relative to the date for each log entry's midnight. In the end though, having to parse the entire file to get ouput kind of defeats the point of using a generatory. Knowing that the log is in descending order, you could have read the file in reverse and stopped once you hit midnight. 

## /users/abrick/pr/231/2647/vw.py

In [ ]:
# Demonstrate the use of a generator that indicates for each event in /etc/httpd/logs/access_log the number of seconds elapsed since the most recent mi$

import datetime

# the file we will be using
log = '/etc/httpd/logs/access_log'

# let's declare a time variable for midnight
midnight = datetime.datetime.strptime("00:00:00", '%H:%M:%S')

# here we use a generator expression to iterate through the lines in our file
# we strip each file for the information we need
# we convert the string data we get into a datetime object
# we subtract midnight from this object and convert this to seconds
times = ((datetime.datetime.strptime(((' '.join(line.split(" ")[3:4]))[13:21]), '%H:%M:%S') - midnight).total_seconds() for line in open(log))

# now we can print data for the number of lines we indicate (i chose 10)
print([next(times) for _ in range(10)])

Peer review: 
    
Really simple solution, excellent parsing of the file lines, good clear comments. A regex might be more flexible, but assuming the file format remains constant, this works wonderfully.

## /users/abrick/pr/231/2647/ed.py

In [38]:
from datetime import datetime

inline_generator = (line.split('[')[1][12:20] for line in open('access_log', 'r'))

now = datetime.now()

for i in range(10):
    print(next(inline_generator))
    print((datetime.strptime(next(inline_generator), "%H:%M:%S") -
           now.replace(hour=0, minute=0, second=0)).total_seconds())

09:36:00
-3810723840.278963
09:36:01
-3810723704.278963
09:38:17
-3810723702.278963
09:38:59
-3810723453.278963
09:42:28
-3810723452.278963
09:42:28
-3810723452.278963
09:42:28
-3810723452.278963
09:42:29
-3810723451.278963
09:42:29
-3810723451.278963
09:44:12
-3810723346.278963


Peer review:

Missing a '/' in the original filepath, causing the program to fail. Once fixed, the output format seems a bit unclear, "-3810723840.278963". I'm not sure where decimals would come into play. That's a faily complicated print statement, the formatting might be more clear if done earlier in the process. 

## /users/abrick/pr/231/2647/et.py

In [1]:
'''
CS231, Homework 5, due 09/28/20
Assignment: Demonstrate the use of a generator that indicates for each event in access_log the number
of seconds elapsed since the most recent midnight.
'''

import datetime

# create a sequence of timestamps with only the date and time using a generator expression
# lines are split by spaces and index 3 is selected, then indexes 1 to end of index 3
log_timestamps = ((line.split()[3])[1:] for line in open('access_log'))

''' Note: can't place above generator assignment statement within generator function below b/c will
    cause a printout of only the first event repeated over the printed range of results. '''

# use a generator function to convert the timestamp generator object sequence to date objects
# recent midnights = each date object with time set to zero
def log_date_objects():
    for i in log_timestamps:
        date_obj = datetime.datetime.strptime(i, '%d/%b/%Y:%H:%M:%S') # converts to a date object
        recent_midnight = date_obj.replace(hour=0, minute=0, second=0, microsecond=0)
        seconds_elapsed = int((date_obj - recent_midnight).total_seconds())
        result_str = "has"
        yield i, result_str, seconds_elapsed

# print selection of sequence: iterate through range of date object generator using a list comprehension
for i in [next(log_date_objects()) for _ in range (15)]:
    print ("Event",*i, "seconds since midnight")


Event 14/Aug/2017:09:36:00 has 34560 seconds since midnight
Event 14/Aug/2017:09:36:00 has 34560 seconds since midnight
Event 14/Aug/2017:09:36:01 has 34561 seconds since midnight
Event 14/Aug/2017:09:38:16 has 34696 seconds since midnight
Event 14/Aug/2017:09:38:17 has 34697 seconds since midnight
Event 14/Aug/2017:09:38:18 has 34698 seconds since midnight
Event 14/Aug/2017:09:38:59 has 34739 seconds since midnight
Event 14/Aug/2017:09:42:27 has 34947 seconds since midnight
Event 14/Aug/2017:09:42:28 has 34948 seconds since midnight
Event 14/Aug/2017:09:42:28 has 34948 seconds since midnight
Event 14/Aug/2017:09:42:28 has 34948 seconds since midnight
Event 14/Aug/2017:09:42:28 has 34948 seconds since midnight
Event 14/Aug/2017:09:42:28 has 34948 seconds since midnight
Event 14/Aug/2017:09:42:28 has 34948 seconds since midnight
Event 14/Aug/2017:09:42:29 has 34949 seconds since midnight


Peer review:

Great docstring, good comments. Really clear and readable output, and really clear syntax. Others have more concise syntax, but this is more reabable. Interesting use of * to unpack in the print statement, I'll remember that one. Your note could probably be removed, seems more of something to know during development but doesn't help in the future (it is possible to put that line in the same single generator), here how I did it: for line in open(file_path): \ timestamp = re.search('(?<=\[).{20}', line).group(0) \ results_dt = datetime.strptime(timestamp, '%d/%b/%Y:%H:%M:%S') \ sec_elapsed = results_dt.hour * 60 * 60 + results_dt.minute * 60 + results_dt.second \ yield timestamp, sec_elapsed

## /users/abrick/pr/231/2647/nw.py

In [ ]:
import datetime

#first, opening the access log with the below code

file=open(r'/etc/httpd/logs/access_log','r')

#then, splitting lines by filtering the 4th item in each line(array)

line= (line.split(" ")[3] for line in file)

# getting a list consisting of first 10 events in the access log.
list = ([next(line) for _ in range(10)])

#loop for the elements in the list
for i in list:
#converting the elements into dates.
        dt = (datetime.datetime.strptime(i, "[%d/%b/%Y:%H:%M:%S"))
#replacing the hour,minute and second with 0 to get the midnight of the related date
        midnight=dt.replace(hour=0, minute=0, second=0)
#calculating the difference and converting into seconds
        timedelta=(dt - midnight).seconds
#result
       	print("event: " + format(dt) + " - seconds elapsed since the midnight " + format(timedelta))

Peer review:

Great and reabable output. Your timedelta solution might be my favorite method for doing this, super clear. Great balance between brevity and redability. One challenge her *might* be the list generator. If reading the whole file, will results be produced before that has been run through the file? I don't know.

## /users/abrick/pr/231/2647/7w.py

In [ ]:
# Demonstrate the use of a generator that indicates for each event in access_log the number of seconds elapsed since the most recent midnight.

from datetime import datetime
import re

log = open(r'/etc/httpd/logs/access_log','r')

# This function pulls out the timestamp from each line in the log, and converts it into a datetime object.
# It also calculates midnight of the timestamp and subtracts timestamp - midnight
def elapsed():
    time_gen = (re.findall(r'\d\d/\w{3}/\d\d\d\d:\d\d:\d\d:\d\d', line) for line in log)
    for timestamp in time_gen:
        time_obj = datetime.strptime(timestamp[0], '%d/%b/%Y:%H:%M:%S')
        midnight_obj = datetime.strptime(str(timestamp[0])[0:11]+':00:00:00', '%d/%b/%Y:%H:%M:%S')
        yield time_obj - midnight_obj

# Printing the next 20 yields in seconds
for i in range(20):
    print(next(elapsed()).seconds)

Peer review:

Looks good, works good, is good. Writing the "long-form" generator makes this much more readable; follows a simple flow that made this one of the easier generators to comprehend.